soweego/linker/classifiers.py

```diff
@@ -37,12 +38,152 @@
+        raise NotImplementedError(
+            'Subclasses need to implement the "create_model" method.'
+        )
+
+    def _predict(self, values):
+        return self.kernel.predict(values)
+
+    def __repr__(self):
+        return (
+            f'{self.__class__.__name__}('
+            f'optimizer={self.optimizer.__class__.__name__}, '
+            f'loss={self.loss}, '
+            f'metrics={self.metrics})'
+        )
+
+
+class _MLensAdapter(SKLearnAdapter, BaseClassifier):
+    """
+    Wrapper around :class:`recordlinkage.SKLearnAdapter`, and
+    :class:`BaseClassifier` to be used as parent class for any classifier
+    which uses as kernel a subclass of :class:`mlens.ensemble.base.BaseEnsemble`.
+
+    This *adapter* correctly implements the *prob* and *_predict* methods so
+    the classifier can be properly used with the *recordlinkage* framework.
+    """
+
+    def __init__(self, **kwargs):
+        super(_MLensAdapter, self).__init__()
+
+    def _check_correct_pred_shape(self, preds: np.ndarray):
+        """
+        Sanity check to see that the *meta* model in the ensemble
+        actually gave as an output two possible classes.
+        """
+        n_classes = preds.shape[1]
+        if n_classes != 2:
+            err_msg = (
+                "We're doing binary classification and we expect "
+                f"probabilities for only two classes, however "
+                f"we received '{n_classes}' classes."
+            )
+            LOGGER.critical(err_msg)
+            raise AssertionError(err_msg)
+
+    def prob(self, feature_vectors: pd.DataFrame) -> pd.Series:
+        """Classify record pairs and include the probability score
+        of being a match.
+
+        :param feature_vectors: a :class:`DataFrame <pandas.DataFrame>`
+          computed via record pairs comparison. This should be
+          :meth:`recordlinkage.Compare.compute` output.
+          See :func:`extract_features() <soweego.linker.workflow.extract_features>`
+          for more details
+        :return: the classification results
+        """
+
+        # mlens `predict` method returns probabilities
+        classifications = self.kernel.predict(feature_vectors)
+        self._check_correct_pred_shape(classifications)
+
+        # we're only interested in the probability for the positive
+        # case
+        classifications = classifications[:, 1]
+
+        return pd.Series(classifications, index=feature_vectors.index)
+
+    def _predict(self, features) -> np.ndarray:
+        prediction = super(_MLensAdapter, self)._predict(features)
+        self._check_correct_pred_shape(prediction)
+
+        prediction = prediction[:, 1]
+
+        # mlens `predict` method returns probabilities. Since we're
+        # dealing with a binary classification problem we just get the
+        # probabilities for the positive case and round them to [0,1].
+        prediction = np.array(list(round(x) for x in prediction))
```

Comment:
Could you explain what's the exact goal of this decision? Does `round them to [0,1]` mean that we're dropping the exact probability scores?

Reply:
Thanks @marfox you raise a good point. However, as you can see I've added two methods: `prob` which returns the raw predictions, and `_predict` which returns the predicted label (that is why I'm using `round`). 

To clarify, `_predict` is used by `recordlinkage` during the evaluation, since it needs the final class labels to calculate the metrics. `prob` is used the rest of the time (in `train` and `link`)